crates/data-plane-controller/tests/util.rs

```diff
@@ -0,0 +1,145 @@
+use data_plane_controller::{
+    commands,
+    controller::{EmitLogFn, RunCmdFn},
+    stack,
+};
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use itertools::Itertools;
+use sqlx::types::uuid;
+use std::sync::{Arc, Mutex};
+
+#[derive(serde::Serialize)]
+pub enum TraceEntry {
+    Log(&'static str, String),
+    Cmd(&'static str, String),
+    Status(stack::Status),
+    Exports(stack::ControlExports),
+    Stack(stack::PulumiStack),
+}
+
+pub fn mock_emit_log_fn(trace: Arc<Mutex<Vec<TraceEntry>>>) -> EmitLogFn {
+    Box::new(
+        move |_logs_token, stream, message| -> BoxFuture<'static, anyhow::Result<()>> {
+            trace.lock().unwrap().push(TraceEntry::Log(stream, message));
+
+            futures::future::ready(Ok(())).boxed()
+        },
+    )
+}
+
+pub fn mock_run_cmd_fn(trace: Arc<Mutex<Vec<TraceEntry>>>) -> RunCmdFn {
```

Comment:
So to write tests that check behaviour of specific states, we would want to extend this function to accept a custom fixture to use instead of `dry_run_fixture` so each test can customise the state it conveys. It would be cool actually to do a merge patch of `dry_run_fixture` with the input so that each test only applies whatever it specifically is testing instead of writing whole fixture files.

I can look into that in my pull-request, but want to make sure we are aligned on that

Reply:
Yea, I have no specific opinion on the nature of `RunCmdFn`'s used in tests and how they're built.

I put this together because I wanted to demonstrate a working and useful garden path leveraging snapshots, and I extracted into `util.rs` on the presumption that we'd want to re-use it, but IMO tests can do whatever they want here.